### DISCUSSION TITLE
Is there a way to send a gratuis ARP with this library?

### DISCUSSION THREAD

**Discussion by ojoojoo** [Networking-for-Arduino/EthernetENC#52] [General]
I'm looking for a way to maintain the ARP table on the router, it now times out after a while. The only solution now is to add a manual ARP entry to the ARP table of the router, but this is unwanted.
I now have this, but it seems to hang somewhere if I add this to my code:

  // Send a Gratuitous ARP
  sendGratuitousARP();

}

 void sendGratuitousARP() {
  // Construct a Gratuitous ARP packet manually
  uint8_t arpPacket[42] = {
    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, // Ethernet destination MAC (broadcast)
    mac[0], mac[1], mac[2], mac[3], mac[4], mac[5], // Ethernet source MAC
    0x08, 0x06, // ARP type (0x0806)
    0x00, 0x01, // Hardware type (Ethernet)
    0x08, 0x00, // Protocol type (IPv4)
    0x06, // Hardware address length
    0x04, // Protocol address length
    0x00, 0x01, // ARP request operation
    mac[0], mac[1], mac[2], mac[3], mac[4], mac[5], // Sender hardware address (MAC)
    192, 168, 1, 100, // Sender protocol address (IP)
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // Target hardware address (zero-filled)
    192, 168, 1, 100  // Target protocol address (IP)
  };

  // Send the packet via UDP to the local IP
  udp.beginPacket(Ethernet.localIP(), 9); // Send to self on port 9 (arbitrary)
  udp.write(arpPacket, sizeof(arpPacket));
  udp.endPacket();
 }

**Comment by ojoojoo** [comment]
This is the entire code:


#include <SPI.h>
#include <EthernetENC.h>
#include <EthernetUdp.h> // Added for sending UDP packet

// Set the MAC address of your Ethernet shield.
byte mac[] = { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xED };

// Define the relay pins
const int relayPinA1 = A1;
const int relayPinA2 = A2;

EthernetUDP udp; // Create an instance of EthernetUDP

void setup() {
  Serial.begin(9600);
  pinMode(relayPinA1, OUTPUT);
  pinMode(relayPinA2, OUTPUT);

  // Initialize the relays to an initial state (off)
  digitalWrite(relayPinA1, LOW);
  digitalWrite(relayPinA2, LOW);

  // Initialize the Ethernet shield
  Ethernet.begin(mac);
  if (Ethernet.hardwareStatus() == EthernetNoHardware) {
    Serial.println("Ethernet module not found. Check your connections.");
    while (true);
  }
  if (Ethernet.linkStatus() == LinkOFF) {
    Serial.println("Ethernet cable is not connected.");
    while (true);
  }
  Serial.println("Ethernet connected");

  udp.begin(9); // Start UDP on arbitrary port (9 in this case)
}

void loop() {
  // Renew DHCP lease and maintain Ethernet connection
  Ethernet.maintain();

  // Send a Gratuitous ARP
  sendGratuitousARP();

}

 void sendGratuitousARP() {
  // Construct a Gratuitous ARP packet manually
  uint8_t arpPacket[42] = {
    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, // Ethernet destination MAC (broadcast)
    mac[0], mac[1], mac[2], mac[3], mac[4], mac[5], // Ethernet source MAC
    0x08, 0x06, // ARP type (0x0806)
    0x00, 0x01, // Hardware type (Ethernet)
    0x08, 0x00, // Protocol type (IPv4)
    0x06, // Hardware address length
    0x04, // Protocol address length
    0x00, 0x01, // ARP request operation
    mac[0], mac[1], mac[2], mac[3], mac[4], mac[5], // Sender hardware address (MAC)
    192, 168, 1, 100, // Sender protocol address (IP)
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, // Target hardware address (zero-filled)
    192, 168, 1, 100  // Target protocol address (IP)
  };

  // Send the packet via UDP to the local IP
  udp.beginPacket(Ethernet.localIP(), 9); // Send to self on port 9 (arbitrary)
  udp.write(arpPacket, sizeof(arpPacket));
  udp.endPacket();
 }

**Reply by JAndrassy** [reply]
you can't send ARP as UDP. the packet type value in the packet's header is different. additinally endPacket will send an ARP request if it doesn't have the MAC address for the target IP address of the UDP message cached.

the core of this library is the old uIP library. it can send arp packets (of course). The Arduino wrapper (Ethernet class & co.) uses ARP internally, but there is no Arduino API to send and receive ARP type packets.

**Comment by JAndrassy** [comment]
I see a problem. The ARP entries never expire now, because the uip_arp_timer() which does the cleanup is never called. I fix it.

**Reply by ojoojoo** [reply]
If you can fix this than I guess my problem is over, it would be great if the arp entry in the router gets renewed by the Arduino. I was trying to make this work with another library, the UIPEthernet, but than it runs into memory issues due to memory fragmentation after just two days. Thanks for the help

**Reply by JAndrassy** [reply]
I don't know if it helps, but now the library will make an ARP request every 20 minutes, if it sends data

**Reply by ojoojoo** [reply]
I will test it out the coming week and let you know the result, thanks so far

**Reply by ojoojoo** [reply]
Hi,

Unfortunately is the Arduino NANO hanging after 2 days, I think it's memory related, so I'm not able to fully test if your ARP modification is working. Maybe if you can find the time, you can look at my code and find the problem?

```
#include <SPI.h>
#include <EthernetENC.h>
#include <EthernetUdp.h>

byte mac[] = { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xED };

const int relayPinA1 = A1;
const int relayPinA2 = A2;

EthernetClient client;
EthernetUDP udpClient;

IPAddress gateway;

unsigned long lastPingTime = 0;
unsigned long lastRelayTransitionTimeA1 = 0;
unsigned long lastRelayTransitionTimeA2 = 0;

enum RelayState {
  RELAY_OFF,
  RELAY_TRANSITION_ON,
  RELAY_ON,
  RELAY_TRANSITION_OFF
};

RelayState relayA1State = RELAY_OFF;
RelayState relayA2State = RELAY_OFF;

void setup() {
  pinMode(relayPinA1, OUTPUT);
  pinMode(relayPinA2, OUTPUT);
  digitalWrite(relayPinA1, HIGH);
  digitalWrite(relayPinA2, HIGH);

  Ethernet.begin(mac);
  if (Ethernet.hardwareStatus() == EthernetNoHardware) {
    while (true);
  }
  if (Ethernet.linkStatus() == LinkOFF) {
    while (true);
  }

  udpClient.begin(8888);
  gateway = Ethernet.gatewayIP();
}

void relayTransition(int relayPin, RelayState &state, unsigned long &lastTransitionTime) {
  switch (state) {
    case RELAY_OFF:
      digitalWrite(relayPin, LOW);
      state = RELAY_TRANSITION_ON;
      lastTransitionTime = millis();
      break;
    case RELAY_TRANSITION_ON:
      if (millis() - lastTransitionTime >= 25000) {
        digitalWrite(relayPin, HIGH);
        state = RELAY_ON;
        lastTransitionTime = millis();
      }
      break;
    case RELAY_ON:
      if (millis() - lastTransitionTime >= 5000) {
        state = RELAY_TRANSITION_OFF;
        lastTransitionTime = millis();
      }
      break;
    case RELAY_TRANSITION_OFF:
      if (millis() - lastTransitionTime >= 25000) {
        digitalWrite(relayPin, HIGH);
        state = RELAY_OFF;
        lastTransitionTime = millis();
      }
      break;
  }
}

bool checkInternetConnection() {
  IPAddress dnsServer(8, 8, 8, 8);
  if (client.connect(dnsServer, 53)) {
    client.stop();
    return true;
  } else {
    return false;
  }
}

void loop() {
  Ethernet.maintain();

  bool isConnected = checkInternetConnection();

  if (!isConnected) {
    relayTransition(relayPinA1, relayA1State, lastRelayTransitionTimeA1);
    
    if (relayA1State != RELAY_TRANSITION_ON && relayA1State != RELAY_TRANSITION_OFF) {
      relayTransition(relayPinA2, relayA2State, lastRelayTransitionTimeA2);
    }
    
    for (int i = 0; i < 5; i++) {
      udpClient.beginPacket(gateway, 32080);
      udpClient.write("Ping");
      udpClient.endPacket();
      delay(1000);
      isConnected = checkInternetConnection();
      if (isConnected) {
        break;
      }
    }
  } else {
    // Clear relay transition states if internet is connected
    relayA1State = RELAY_OFF;
    relayA2State = RELAY_OFF;

    digitalWrite(relayPinA1, HIGH); // Set relayPinA1 to HIGH
    digitalWrite(relayPinA2, HIGH); // Set relayPinA2 to HIGH

    unsigned long currentTime = millis();
    if (currentTime - lastPingTime >= 10000) {
      lastPingTime = currentTime;
      udpClient.beginPacket(gateway, 32080);
      udpClient.write("Ping");
      udpClient.endPacket();
    }
  }

  delay(10); // A small delay to avoid busy-waiting
}
```
It would be much appreciated,

Orson

**Reply by JAndrassy** [reply]
try to monitor the memory
https://github.com/mpflaga/Arduino-MemoryFree

**Reply by ojoojoo** [reply]
Thanks for the help, I don't think it's the memory, because the amount of memory remains the same:
14:48:11.530 -> Old way to force String to Flash
14:48:11.578 -> New way to force String to Flash
14:48:11.578 -> Free RAM = 
14:48:11.578 -> 812
14:48:11.720 -> Old way to force String to Flash
14:48:11.768 -> New way to force String to Flash
14:48:11.815 -> Free RAM = 
14:48:11.815 -> 812


But suddenly it just hangs, and is unresponsive? Can you please point me in the right direction on how to troubleshoot this problem?

**Reply by JAndrassy** [reply]
can you try it without the UDP pings?

**Reply by ojoojoo** [reply]
I let it run for the entire weekend, but unfortunately also with UDP ping disabled I have the exact same problem, it just hangs for no reason.
It has enough free memory, that doesn't change at all.

15:28:58.800 -> Old way to force String to Flash
15:28:58.848 -> New way to force String to Flash
15:28:58.848 -> Free RAM = 
15:28:58.848 -> 812

This is the code I was using:

#include <SPI.h>
#include <EthernetENC.h>
#include <EthernetUdp.h>
#include <MemoryFree.h>;
#include <pgmStrToRAM.h>; // not needed for new way. but good to have for reference.

byte mac[] = { 0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xED };

const int relayPinA1 = A1;
const int relayPinA2 = A2;

EthernetClient client;
EthernetUDP udpClient;

IPAddress gateway;

unsigned long lastPingTime = 0;
unsigned long lastRelayTransitionTimeA1 = 0;
unsigned long lastRelayTransitionTimeA2 = 0;

enum RelayState {
  RELAY_OFF,
  RELAY_TRANSITION_ON,
  RELAY_ON,
  RELAY_TRANSITION_OFF
};

RelayState relayA1State = RELAY_OFF;
RelayState relayA2State = RELAY_OFF;

void setup() {
  pinMode(relayPinA1, OUTPUT);
  pinMode(relayPinA2, OUTPUT);
  digitalWrite(relayPinA1, HIGH);
  digitalWrite(relayPinA2, HIGH);

  Ethernet.begin(mac);
  if (Ethernet.hardwareStatus() == EthernetNoHardware) {
    while (true);
  }
  if (Ethernet.linkStatus() == LinkOFF) {
    while (true);
  }

  udpClient.begin(8888);
  gateway = Ethernet.gatewayIP();
}

void relayTransition(int relayPin, RelayState &state, unsigned long &lastTransitionTime) {
  switch (state) {
    case RELAY_OFF:
      digitalWrite(relayPin, LOW);
      state = RELAY_TRANSITION_ON;
      lastTransitionTime = millis();
      break;
    case RELAY_TRANSITION_ON:
      if (millis() - lastTransitionTime >= 5000) {
        digitalWrite(relayPin, HIGH);
        state = RELAY_ON;
        lastTransitionTime = millis();
      }
      break;
    case RELAY_ON:
      if (millis() - lastTransitionTime >= 600000) {
        state = RELAY_TRANSITION_OFF;
        lastTransitionTime = millis();
      }
      break;
    case RELAY_TRANSITION_OFF:
      if (millis() - lastTransitionTime >= 5000) {
        digitalWrite(relayPin, HIGH);
        state = RELAY_OFF;
        lastTransitionTime = millis();
      }
      break;
  }
}

bool checkInternetConnection() {
  IPAddress dnsServer(8, 8, 8, 8);
  if (client.connect(dnsServer, 53)) {
    client.stop();
    return true;
  } else {
    return false;
  }
}

void loop() {
  Ethernet.maintain();

  bool isConnected = checkInternetConnection();

  if (!isConnected) {
    relayTransition(relayPinA1, relayA1State, lastRelayTransitionTimeA1);
    
    if (relayA1State != RELAY_TRANSITION_ON && relayA1State != RELAY_TRANSITION_OFF) {
      relayTransition(relayPinA2, relayA2State, lastRelayTransitionTimeA2);
    }
    
    for (int i = 0; i < 5; i++) {
//      udpClient.beginPacket(gateway, 32080);
//      udpClient.write("Ping");
//      udpClient.endPacket();
      delay(1000);
      isConnected = checkInternetConnection();
      if (isConnected) {
        break;
      }
    }
  } else {
    // Clear relay transition states if internet is connected
    relayA1State = RELAY_OFF;
    relayA2State = RELAY_OFF;

    digitalWrite(relayPinA1, HIGH); // Set relayPinA1 to HIGH
    digitalWrite(relayPinA2, HIGH); // Set relayPinA2 to HIGH

    unsigned long currentTime = millis();
    if (currentTime - lastPingTime >= 10000) {
      lastPingTime = currentTime;
//      udpClient.beginPacket(gateway, 32080);
//      udpClient.write("Ping");
//      udpClient.endPacket();
    }
  }
  // put your setup code here, to run once:
  Serial.begin(9600);
  Serial.println(getPSTR("Old way to force String to Flash")); // forced to be compiled into and read   
  Serial.println(F("New way to force String to Flash")); // forced to be compiled into and read   
  Serial.println(F("Free RAM = ")); //F function does the same and is now a built in library, in IDE > 1.0.0
  Serial.println(freeMemory());  // print how much RAM is available in bytes.
  // print how much RAM is available.
  delay(10); // A small delay to avoid busy-waiting
}

**Comment by JAndrassy** [comment]
a week is good. I didn't have a better result and I could not find out what is the problem..  
use watchdog to reset the sketch when it hangs.
the W5500 with Ethernet library is not better.
